docs/architecture/Data_caching_lineage.md

```diff
@@ -0,0 +1,297 @@
+
+Second case is a situation when an actual set of file to use is a significantly smaller part of the dataset. It's very typical for specific experiment pipelines or integration tests. While it may optimize premium storage utilization and caching time, maintaining lists of files to copy requires additional investment, such as annotations processing. For example, if annotations are stored in Pascal VOC format, there may be a code component that builds a list of labeled images so that only these files are cached.
+
+### Recommended Data Caching Strategy
+
+If images are mostly shared between pipelines and experiments in a single cache storage, caching the whole dataset is an optimal choice as it doesn't bring additional complication to data management.
+
+## Data Lineage in Azure Machine Learning
+
+**Model reproducibility** is a critical component of production-grade machine learning pipelines. Without it, Data scientists risk having limited visibility into what causes changes in model performance. The variability may appear to be caused by an adjustment of one parameter, but may actually be caused by hidden sources of randomness. Reproducibility reduces or eliminates variations when rerunning failed jobs or prior experiments, making it essential in the context of fault tolerance and iterative refinement of models. Running ML pipelines in the cloud across multiple compute nodes not only multiplies the sources of non-determinism but also increases the need for both fault tolerance and iterative model development.
+
+There are 5 major components that define a machine learning model:
+
+1. Model architecture
+1. Model hyperparameters
+1. Training code
+1. Data augmentation code and parameters
+1. Data and annotations
+
+First 4 components are tracked between pipeline's Git repository and Azure Machine Learning tracking storages. Therefore, it's relatively easy to achieve model reproducibility up to the point of what data was used.
+
+Data and annotations are much harder to track. There are tools and processes to maintain data versioning that can provide a great level of observability on data. However, with Azure Machine Learning pipelines, the same level of observability may be achieved with significantly lower effort. For example, here's what we can do when training computer vision models.
+
+1. Maintain snapshots of annotation files as they are expected to change. Every pipeline run has a snapshot of annotation files in the run's output storage.
+1. Maintain a single shared cache of data files (images) as they are expected to stay unmodified once added to the storage.
+1. Model training code uses dataset annotation data to get references to data files (images). As data files are added over time, previous experiments are still fully reproducible based on the stored annotation snapshots.
+
+### Recommended Data Lineage Strategy
+
+For computer vision models trained on Azure Machine Learning pipelines, we recommend the following strategy:
```

Comment:
How does this account for image preprocessing tasks that may only need to run once on a given set of images? In practice, we typically want to keep the raw images, and the preprocessed ones with some metadata that allows us to know if the preprocessing logic has changed, and then re-preprocess when the preprocessing logic has been altered. Then in training we would only do the image processing required for the actual training process.

Reply:
Good call. Will let @vladkol answer this question.

Reply:
The strategy requires some context setting. We had frame extraction as a separate step performed in Azure Data Factory. Recorded videos were split to frames, frames were stored in long-term storage. After that, recorded videos were moved to the archival storage, operated separately and under a different policy.

Reply:
Applied Vlad's comment.